### ISSUE TITLE
Improve 3.1+ schema `$id` UX

### ISSUE THREAD

**Issue by handrews** [OAI/OpenAPI-Specification#4147]
Currently, we have the following schema document `$id`s for OAS 3.1 (shown relative to a base URI of `https://https://spec.openapis.org/oas/3.1/`, and without the trailing date segment that all schemas need to support updates:

* `schema`   # Validates everything _except_ the Schema Object, which is only validated against `type: [boolean, object]`
* `schema-base`  # Extends `schema` to validate Schema Objects against `dialect/base` _and_ require both `$schema` and `jsonSchemaDialect` to be set to `schema`'s `$id` if they are present
* `dialect/base` # A dialect equivalent to the default 2020-12 dialect plus the OAS extensions
* `meta/base`  # The vocabulary metaschema for the OAS extensions

`schema-base` was named to indicate that it incorporates `dialect/base`.  No one is certain why the dialect and vocabulary are referred to as "base", but they are called that in the OAS text as well.  I vaguely recall discussing the dialect as a base for further OAS-relevant extensions?  Perhaps that is what the intent was.

AFAICT, the advantage of `schema` is that it's self-contained.  But giving it that name means that it is the schema that most people assume is what they want, yet it really does not produce the behavior they'd expect.

`schema-base` is closer to what people would expect, but is (in my view) overly restrictive in locking `$schema` to `dialect/base`, as it is valid to switch metaschemas if `$schema` appears alongside `$id`, and conforming JSON Schema implementations know to treat `$id`-containing subschemas as embedded resources and check for a changed metaschema.  However, this does ***not*** make `schema-base` wrong:  Users might indeed want to enforce using _only_ `dialect/base`, and this is the correct way to do that.

On the other hand (and contrary to some of my comments earlier on Slack), `schema-base` locking `jsonSchemaDialect` to `dialect/base` is probably correct.  If you want to set `jsonSchemaDialect` to something else, you would want a schema that validates Scheam Objects accordingly by default.

-----

We probably shouldn't change how `schema` behaves, but we should make it more obvious that it is incomplete.  We should also come up with some sort of intuitive name for a schema that is like `schema-base` but does not restrict `$schema`.  This name needs to convey that _most users will want this new schema, and **not** the plain_ `schema`.

Ideally, I'd prefer to make copies of the vocabulary and dialect and give them the URIs `dialect/oas` and `meta/oas` (`3.1` is already present earlier in the URI), but that is probably more trouble than it is worth given that I imagine very few people pay attention to those other than to just load them correctly when using `schema-base`.

The important thing is to have an intuitively-named schema that performs the default accepted Schema Object validation but still allows switching metaschemas on embedded resources.  That is the default behavior of the OAS, and it's not currently provided.

**Comment by fpapon** [comment]
Hi,

I'm sure to understand which url to use to validate the 3.1.0 and the 3.1.1 version of the specification.
Why using the date in the url instead of the last version number?

https://spec.openapis.org/oas/3.0/schema/2024-10-18 => https://spec.openapis.org/oas/3.0.0/schema

I am asking because in the openapi.json files, we are using the full version number so it's easier to find the right jsonschema to use.

**Comment by handrews** [comment]
@fpapon There are no schema-visible differences between patch releases.  That is part of the _definition_ of a patch release.  The dates allow us to make bug fixes and other improvements.  But those have nothing to do with patch releases.  All 3.0 schemas apply to 3.0.0, 3.0.1, 3.0.2, 3.0.3, 3.0.4, etc., and all 3.1 schemas apply to 3.1.0, 3.1.1, 3.1.2 (forthcoming), etc. but the most recently dated one is the most accurate.

Again, there is _no difference_ among patch releases that can be reflected in a schema.

**Comment by handrews** [comment]
@fpapon but see also #4877.

**Comment by rmannibucau** [comment]
@handrews while I fully get what you mean I think François has two points:

- there is no real point in not releasing the schema with the spec so at some point the versioning should be 1-1 and it will help end users to not have to fight to find the schema (it is "common" and "natural")
- the date in the current url is misleading and implies a double versioning unrelated to versions so it is very awckward even if I get from where it comes

So overall I think it should be fixed, even if it is just to get a better `"documentation"` wording in another patch release

**Comment by ralfhandl** [comment]
@rmannibucau 

> there is no real point in not releasing the schema with the spec so at some point the versioning should be 1-1

We actually have _eight_ iterations of the v3.1 schemas on https://spec.openapis.org/#openapi-specification-schemas while the v3.1 spec only has reached patch version 1, on average four schema patches per spec patch.

Once we publish spec v3.1.2 and a subsequent schema iteration, the ratio will drop to 3-1, still way off 1-1.

> the date in the current url is misleading and implies a double versioning unrelated to versions

Schema fixes ("iterations") can and do happen more often than spec fixes ("patch versions"), so they are versioned differently.

We could have used sequence numbers for the iterations instead of dates, so the schemas would be `3.1-0`, ..., `3.1-7`. Not sure whether that would be less misleading.

**Comment by rmannibucau** [comment]
@ralfhandl 

> We actually have eight iterations of the v3.1 schemas on https://spec.openapis.org/#openapi-specification-schemas while the v3.1 spec only has reached patch version 1, on average four schema patches per spec patch.

which is an issue

it means specifying a spec version doesn't target a schema and a schema doesn't align on a spec version.

By itself no issue until you can only specify either the spec version or the schema iteration in the spec document which is the case today.

I clearly vote to not enable to specify both but to just do a minor with everything every time it happens, it is not a crazy work and will help everyone.

Side note: if the answer is "you set the version and the $schema" this is just not true cause schema are often injected or handled transversally so we literally have only the spec version to deduce the schema to use and changes like uri/uri-reference are impacting, even in patch releases and totally missed today (jsonSchemaDialect as an example).


So long story short: I think schema should be versionned 1-1 with the spec since they are part of it from an user (and integrator) standpoint.

**Comment by fpapon** [comment]
Hi,
My concern was, in a openapi.json we can have:

```
{
  "openapi": "3.1.0"
}
```

which should match on 1-1 version of the jsonSchema, this is used by editors to automatically run the json validation without having the jsonSchema referenced in the document.

```
{
  "$schema": "https://spec.openapis.org/oas/3.1/schema/2025-02-13"
  "openapi": "3.1.0"
}
```

So here, I don't know which jsonSchema matched the 3.1.0 version. As I understand on this thread, the latest 3.1 jsonSchema should match all 3.1.x version but why not having just one version 3.1 or follow the incremental patch version?

I saw that the rule of matching 3.1.x in the jsonSchema checking that compatibility:

```
    "openapi": {
      "type": "string",
      "pattern": "^3\\.1\\.\\d+(-.+)?$"
    },
```

So maybe publishing only the latest spec should be enough and easier to use:

```
{
  "$schema": "https://spec.openapis.org/oas/3.1/schema
  "openapi": "3.1.0"
}
```

**Comment by ralfhandl** [comment]
@rmannibucau 

> I think schema should be versionned 1-1 with the spec since they are part of it from an user (and integrator) standpoint.

That would be a possibility going forward.

It would mean that all schema fixes after publication of a spec patch version are "parked" in the corresponding spec minor-version dev branch (for example `v3.1-dev`) until a new patch version for that minor version is published.

**Comment by ralfhandl** [comment]
@fpapon Not sure what the JSON examples should convey, `$id` is not a valid field in the root object of an OpenAPI Document.

> publishing only the latest spec should be enough

We do not want to change a published schema „in place“, so a new schema iteration needs a new retrieval URI and a new `$id` value.

We would like to offer a convenience URI for each spec minor version that _redirects_ (responds with HTTP `307 Temporary Redirect`) to the latest published iteration, see
* #4152

**Comment by rmannibucau** [comment]
think std publication is fine:

* 3.x.y/...
* 3.x/...

**Comment by fpapon** [comment]
> [@fpapon](https://github.com/fpapon) Not sure what the JSON examples should convey, `$id` is not a valid field in the root object of an OpenAPI Document.

Sorry I fixed my comment, it's `$schema`.

> 
> > publishing only the latest spec should be enough
> 
> We do not want to change a published schema „in place“, so a new schema iteration needs a new retrieval URI and a new `$id` value.
> 
> We would like to offer a convenience URI for each spec minor version that _redirects_ (responds with HTTP `307 Temporary Redirect`) to the latest published iteration, see
> 
>     * [Make the "latest" schema accessible programmatically #4152](https://github.com/OAI/OpenAPI-Specification/issues/4152)

**Comment by ralfhandl** [comment]
> Sorry I fixed my comment, it's `$schema`.

Well, `$schema` isn't allowed either next to `openapi` in the document root.

I know that some editors/IDEs use this to reference a JSON Schema file for validation, but OpenAPI doesn't support this convention.

**Comment by rmannibucau** [comment]
@ralfhandl i get what you mean but it also means openapi doesn't have any schema since it is a requirement even if almost all tools support other configurations for convenience -> https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-00#rfc.section.8.1.1 , anyway this part is less important since all tools will tolerate this additional keyword

**Comment by handrews** [comment]
@rmannibucau the JSON Schema section that you linked to has nothing to do with using `$schema` in non-JSON Schema documents, which has _never_ been part of the JSON Schema specification.

**Comment by handrews** [comment]
@rmannibucau also, there is no need for such a usage of `$schema` because the `openapi` field gives you the major and minor version, which is sufficient to locate the most up-to-date schema.

**Comment by karenetheridge** [comment]
This confusion with dates and versions has been a concern of mine as well. It's exacerbated by the fact that schemas aren't released _at the same time_ as a specification release - historically they have come weeks or months later, even though the specification isn't generally usable without having the schemas in hand (implementations are often based around the schemas, and users will want to have the new schemas as well to make sure they're correctly using any new features or fixes).

Recognizing that point releases of the specification are meant to be fully backwards-compatible, and therefore the latest version of the schemas in the (for example) 3.1 series should be fully usable with any 3.1.x implementation, this is not actually true in reality because the reverse is not true: **OpenAPI documents written for a particular point release or schema version are not backwards compatible with earlier schemas in the series**. `jsonSchemaDialect` is an example of this: it is now permitted to be a uri-reference, not a uri, so such a uri-reference in an OAD written with the latest schema in hand will be fine, but it cannot be validated with earlier schemas in the 3.1 series; consequently implementations using an earlier version of the schema will not accept this OAD.

This is not a problem that the user can mitigate in any way, nor signal to the implementation what minimum schema version it is expecting, because (as just described above, while I was writing this long reply), the only such signals are:

* `openapi`: indicates the version of the OpenAPI _specification_ that it is targeting (which as said above, can span many different schema iterations);
* `jsonSchemaDialect`: indicates the identifier of the _JSON Schema_ version to use for embedded JSON Schemas (which is not helpful for versioning of the OAD as a whole).

However, if specification releases and schema releases were always done together, then the user could correctly indicate in their OAD with the `openapi` field the information: "this document is expecting the implementation to understand the syntax defined by OAS version 3.1.5, for both specification and schemas". Then the implementation can at least provide a helpful error that this version is beyond its understanding, or perhaps attempt to download new versions from the openapis.org website.

_(And if this doesn't convince you, ask yourself: If indeed point releases are truly insignificant, then why is the `openapi` field specified to always include the point release, rather than ONLY EVER containing the major.minor version? It's not even optional to include the point version -- it's MANDATORY!)_

I find it quite frustrating that there is resistance to multiple instances of users coming forward indicating that _this is actually a real problem in the real world_, and being rebuffed because of a belief that "it _shouldn't_ matter, therefore it _doesn't_ matter".

**Comment by handrews** [comment]
> jsonSchemaDialect is an example of this: it is now permitted to be a uri-reference, not a uri, so such a uri-reference in an OAD written with the latest schema in hand will be fine, but it cannot be validated with earlier schemas in the 3.1 series; consequently implementations using an earlier version of the schema will not accept this OAD.

I don't see why implementations shouldn't let you say which schema to use if you want to manage it on that level.

> why is the openapi field specified to always include the point release, rather than ONLY EVER containing the major.minor version?

It shouldn't in my view, but that discussion stalemated and we don't have a proper decision on it so we're continuing by inertia, which bothers me.  See #4233.

We continuously tell people that it doesn't matter and then ask them to specify it.  I think that is the root of so many UX problems with OpenAPI and I wish we'd fix it.

[I'm moslty not trying to argue what the schema policy _should_ be, just explaining what it _is_, and I wanted to comment on the patch release in the `openapi` field thing because it really bothers me.]

**Comment by handrews** [comment]
@ralfhandl 

> We do not want to change a published schema „in place“, so a new schema iteration needs a new retrieval URI and a new $id value.

Yeah this is where I'm confused.  If we had per-patch-release schemas we'd need to have a whole line of dated schemas _for each patch release_, with the schemas with the same date being identical across all patch releases.  What does that get us?  How is that not _more_ confusing?

Alternatively, would we really want eight patch releases just for schema fixes, which is the only way I can think of to avoid that problem?

Really, I think the fundamental problem is putting the patch release in the `openapi` field.  My vote is still to fix that instead.

**Comment by mehulchauhan** [comment]
what is alternative to
# yaml-language-server: $schema=https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/schemas/v3.1/schema.json

**Comment by rmannibucau** [comment]
@handrews you mix things up, and yes $schema is the standard way to référence a schema - on the phone, maybe I mixed links but it is like that since 10+ years. Yes openapi does have pr does not have to référence it depending the tooling, this is why I said having it or not in the schema is not a topic. But also to référence a schema automatically (tool) or implicitly + deterministically you must either be explicit - where $schema topic comes from - or implicitly, ie with openapi version.
When do you think the new "all in one release" approach can be available please?

**Comment by handrews** [comment]
@rmannibucau I wrote/maintained  musch of the JSON Schema specification.  I am quite certain that `$schema` to reference a schema from an instance is _not_ part of the JSON Schema specification.

**Comment by rmannibucau** [comment]
@handrews yep, doesn't change any of the points or the outcome, does it?
Do you need help to set it up to make it Moving forward?
Happy to help if needed

**Comment by karenetheridge** [comment]
> I don't see why implementations shouldn't let you say which schema to use if you want to manage it on that level.

Assuming you mean "..but not in the OAD itself":

- that forces this information to be sent out-of-band somehow (a configuration parameter to the application? a change in code somewhere?), which is an extra burden on the user
- the user has to know what schema they need to use, which is hard enough for us insiders that I think it would be nearly impossible to the point of incredible frustration for any user, because (_drum roll_) there are no unique versions assigned to schema releases and no release announcements made for them. This forces users to download every iteration of the timestamped schemas to figure out which one actually works, and then write down "this date is required for this change, and that date is required for that change".

All of this would go away if the user could just say _in the document_ "this document was written with release 3.1.25 in mind" and then it's implicit that if only schema version 3.1.24 is available, it might not work. And we already have a field for that, and it doesn't even need changing: that's the `openapi` field at the very top level of the document (https://spec.openapis.org/oas/latest#fixed-fields).

(In my implementation, how I'd handle it is to throw an error explaining clearly that the user asked for a version of the schema that I don't have bundled, which is a signal to the user that either a) they have an old version of the implementation and they need to upgrade, or b) they should file a bug report because the author (moi) hasn't pushed a new release yet that bundles the latest schemas.)

> Yeah this is where I'm confused. If we had per-patch-release schemas we'd need to have a whole line of dated schemas for each patch release, with the schemas with the same date being identical across all patch releases. What does that get us? How is that not more confusing?

Because: 1. the version can actually appear _in the OAD itself_, and 2. there would be a release announcement for it, allowing us to communicate the potentially impactful changes.

> Alternatively, would we really want eight patch releases just for schema fixes?

So what? What's wrong with having a series of 3.1.1 through 3.1.8, each with their own small release notes on a 3.1 changelog page?  There's nothing wrong with iterating quickly -- and it demonstrates to the outside world that something is still happening and the OpenAPI project is still alive.  We like automation, right? If we're already up to pushing point releases like 3.1.1 with minor clarifications in them, why can't we do that for schema changes too?  If it's a tiny fix, I don't think it would be a big deal to hold back the release so the cadence is no more than monthly, but since we love automation, the release process is not a huge hassle, right? Right?  (_padme and anakin_)

**Comment by karenetheridge** [comment]
@mehulchauhan: you should use the schemas as identified here: https://spec.openapis.org/#openapi-specification-schemas -- specifically, the URIs found in the `$id` property in each of those files.  Whatever OpenAPI tool you are using should have these URIs documented and ideally bundle the files themselves right into the application so they don't need to be downloaded from the network.

**Comment by rmannibucau** [comment]
If I summarize facts:

1. Schema are either deduced from the content (implicitly) or explicit ($schema in most editors, even in yaml but in a comment) -> issue is there is not 1-1 relationship between an openapi version and schema (see 2.)
2. Determinism and compatible between schema révision is not guaranteed today (even if desired but it already broke) -> need to link explicitly either one release with a schema. Note that "latest" link matches only one not enterprise friendly case - same you pinpoint docker images at sha level for ex.


So easiest for everyone is to release schema with openapi release and do a new one when at least one of both change.
Alternative - enable to explicit the schema and require it etc... - are not as nice for end users.

Can we work towards that goal?

**Comment by ralfhandl** [comment]
@rmannibucau 

> compatible between schema révision is not guaranteed today (even if desired but it already broke)

With which schema iteration did that happen?

And what changed incompatibly?

**Comment by rmannibucau** [comment]
> Side note: if the answer is "you set the version and the $schema" this is just not true cause schema are often injected or handled transversally so we literally have only the spec version to deduce the schema to use and changes like uri/uri-reference are impacting, even in patch releases and totally missed today (jsonSchemaDialect as an example).
> 

between 3.1.0 and 3.1.1 IIRC - didn't dig more to see if there was others.
The trick there is that it looks like opening so backward compat but it is not true since a working 3.1.0 impl will totally break with 3.1.1. cause it doesn't support relative refs

**Comment by karenetheridge** [comment]
>  The trick there is that it looks like opening so backward compat but it is not true since **a working 3.1.0 impl will totally break with 3.1.1. cause it doesn't support relative refs**

Yes exactly -- except the schemas that make these changes aren't synchronized with 3.1.1, but rather came after 3.1.1 and before 3.1.2 (which hasn't been released yet).

**Comment by handrews** [comment]
_[EDIT: discussion of the real-world `jsonSchemaDialect` example ended up [elsewhere](https://github.com/OAI/OpenAPI-Specification/discussions/4233#discussioncomment-14334060) without me quite realizing it was in the wrong place.  TL;DR: I see it as supporting the current policy, see also [this later comment in this issue](https://github.com/OAI/OpenAPI-Specification/issues/4147#issuecomment-3264241000).]_

> so we literally have only the spec version to deduce the schema to use

There is no deducing.  The schema to use is the latest schema for that minor release line.  If you have a choice, it is _always_ the latest on the minor release line.  The only reason the schemas with earlier dates are left published rather than taken down (and why we put dates in the `$id` so that they _can_ be left published) is because it's possible that an error in an older schema could theoretically _incorrectly allow_ an OAD with an error, and someone may depend on that error.  So we keep them published so folks can use them until they can fix their OADs and use the correct (latest) schema.

>  except the schemas that make these changes aren't synchronized with 3.1.1, but rather came after 3.1.1 and before 3.1.2 (which hasn't been released yet).

There is no "synchronizing".  3.1.2 _replaces_ 3.1.1, which _replaces_ 3.1.0.  They are not distinct versions the way 3.1.x is distinct from 3.0.x.  It's like HTTP RFCs.  While an HTTP implementation _does_ need to know whether to use HTTP/1.0, HTTP/1.1, HTTP/2, or HTTP/3 (distinct versions), if it is implementing HTTP/1.1 _does not_ care about RFC2068 vs RFC2616 vs RFC7230 (and friends) vs RFC9112 (and friends).  Those RFCs replace each other.  All HTTP/1.1 implementations are expected to follow RFC9112, even if they were first written at the time of RFC2068.  There is no way in HTTP to say "I'm using the RFC2616 version of HTTP/1.1."  That's not a thing.  OAS point releases are the same.

(I am aware of some RFCs that obsolete older RFCs in problematic ways, and there may even be examples in the HTTP RFCs, but no system is perfect and that is how it is intended to work.  If your HTTP client connects to a server that talks HTTP/1.1, there is no way to tell or set the RFC involved.)

**Comment by karenetheridge** [comment]
> There is no deducing. The schema to use is the latest schema for that minor release line.

This is another case of theory bumping up against reality. "The schema to use is the latest schema for the release line" sounds simple, but it isn't. If the implementation doesn't have the latest schemas built into it (or rather, it doesn't have a recent enough schema to properly interpret the OAD that the user has provided), _things break_. This is not theory. This is reality. The user cannot simply cause the implementation into having the latest schema through wishing or praying. This is something that needs to be set up in advance.

I have tried, multiple times, to describe a real world example where this can happen, and how it can be made better. How can I explain it better?

**Comment by karenetheridge** [comment]
I find all this especially hilarious since the example that I and @rmannibucau both used is regarding a change to the schemas that haven't even been released yet, even though the changes were done in git _in February_.

**Comment by handrews** [comment]
> I have tried, multiple times, to describe a real world example

I addressed it on [what I now see is a different discussion](https://github.com/OAI/OpenAPI-Specification/discussions/4233#discussioncomment-14334060), I did not realize how the issues/discussions had crossed.

>  is regarding a change to the schemas that haven't even been released yet, even though the changes were done in git in February.

I have no idea why we have not been releasing schemas, I don't pay attention to those PRs.

**Comment by handrews** [comment]
> The user cannot simply cause the implementation into having the latest schema through wishing or praying. This is something that needs to be set up in advance.

But stapling the schema to the point release doesn't help this.  It actually makes it _more_ complicated, and encourages incorrect behavior (using outdated schemas because they appeared to align with a point release when really that alignment is meaningless).

I am _not_ denying the real-world challenges here.  I am just saying that the proposed solution won't solve them, and will in fact cause more problems.  Because we would then need dated schemas for every point release because those later schemas apply to earlier point releases.  Continuing to use an old schema because it was published around the time of an old point release is incorrect.  Changing the way the `$id` is written doesn't make it correct.  It's still the wrong schema.

The solution we need is:

* Stop putting the patch release in `openapi` because it incorrectly makes people think the patch release changes behavior.  This is the fundamental problem.
* Move to a web platform that lets us properly implement a `latest` redirect, or find some other way to make it programmatically accessible.
* Actually publish the npm package to distribute the schemas (there's an issue for this somewhere) so that people can be notified with proper infrastructure that they need to update their dependency.  And maybe packages in other languages as well.

**Comment by handrews** [comment]
@karenetheridge let's go a bit deeper on the `jsonSchemaDialect` example.  If we had distinct schemas for 3.1.0 (with `format: uri` for that field) and 3.1.1 (with `format: uri-reference`), then people whose OADs have `openapi: 3.1.0` would think that they cannot use a relative URI-reference because the schema doesn't allow it.

But that's incorrect.  OAS 3.1.0 _does_ allow relative URI-references for `jsonSchemaDialect`.  So if there was just one schema for 3.1.0 and one for 3.1.1, what would the 3.1.0 user be expected to do?

In theory, they could just change `openapi: 3.1.0` to `openapi: 3.1.1`.  If we think we can get people to do that, then it's worth considering.  But @lornajane 's point on #4233 seems to be that people _won't_ change the `openapi` field, and have some reason not to.  I don't understand this, but Lorna sees a lot more real-world use than I do.

If people won't change `openapi` to get "new" schemas, then tying schemas to point releases isn't a viable solution.

**Comment by rmannibucau** [comment]
> There is no deducing. The schema to use is the latest schema for that minor release line.

as explained this is not an option for several pipeline, it is literally equivalent to say "do not use 3.1.1 but 'latest'" with the exact same impacts so I understand it is current state but we must find a way to make it more robust.

> then people whose OADs have openapi: 3.1.0 would think that they cannot use a relative URI-reference because the schema doesn't allow it

exactly and since it had been the case (previous release) it must be considered as a fact - releases should be seen as immutable.
new release can fix it - and should have been a 3.2.0 from a semver standpoint.

> what would the 3.1.0 user be expected to do

upgrade `openapi` field and get the new schema. Same if you have a bug in a lib, you upgrade it to get a fix. See `openapi` field as your `package.json`/`pom.xml`/...

> If people won't change openapi to get "new" schemas, then tying schemas to point releases isn't a viable solution.

if they do not then they are happy with their current release, all fine for everybody

**Comment by ralfhandl** [comment]
> a working 3.1.0 impl will totally break with 3.1.1. cause it doesn't support relative refs

The source of truth is the spec document, and relative URI references were always allowed by the spec.

An implementation that doesn't support relative URI references is incomplete.

The schema files are _informational_ only. If they are overly restrictive, that's a bug that needs to be fixed, as it was in this case.

**Comment by ralfhandl** [comment]
> the example that I and [@rmannibucau](https://github.com/rmannibucau) both used is regarding a change to the schemas that haven't even been released yet

The schema bug fix for relative URI references was published in Februrary.

The only pending schema bug fixes are
* `server` field of [Link Object](https://spec.openapis.org/oas/v3.1.1.html#fixed-fields-16) was "misspelled" as `body` (a "breaking" change in the schema files)
* `example` and `examples` are mutally exclusive (another "breaking" change in the schema files)

I've approved these changes now, and as soon as the PR gets a second approval, it can be merged and the schema files will be published.

**Comment by rmannibucau** [comment]
> The schema files are informational only. If they are overly restrictive, that's a bug that needs to be fixed, as it was in this case.

Ok, this doesn't change the root issue (explained by @fpapon in https://github.com/OAI/OpenAPI-Specification/issues/4147#issuecomment-3248848048) which is that with such a way you can't link an openapi to a schema just by having the openapi doc so deploying at proper url would be nicer for both tools and end users.
Something like https://spec.openapis.org/oas/x.y.z/schema/jsonschema.json or alike for ex.

Think we agree/converge on the rest.

**Comment by ralfhandl** [comment]
> deploying at proper url would be nicer for both tools and end users.
> Something like https://spec.openapis.org/oas/x.y.z/schema/jsonschema.json or alike for ex.

The "proper URL" wouldn't contain the patch version number `x.y.z` of the spec, it would only contain the minor version number `x.y`.

> which url to use to validate the 3.1.0 and the 3.1.1 version of the specification

The same URL for both 3.1.0 and 3.1.1 because the spec patch version doesn't matter, see [4.1 Versions](https://spec.openapis.org/oas/v3.1.1.html#versions):

> The patch version SHOULD NOT be considered by tooling, making no distinction between `3.1.0` and ` 3.1.1` for example.

**Comment by rmannibucau** [comment]
> The same URL for both 3.1.0 and 3.1.1 because the spec patch version doesn't matter.

I understand the will and know the semver philosophy but this is never true and it costs almost nothing to have both so can you please do `x.y` and `x.y.z`.
This way it is really automatic and more important *deterministic* (`x.y` is not since not immutable)

**Comment by fpapon** [comment]
As I understand, users and tooling should not use full version in their openapi.json file but just the `major.minor` one, but in the jsonschema, the pattern for the openapi version field want le full version.

**Comment by ralfhandl** [comment]
@fpapon unfortunately the spec requires that the `openapi` field uses a `major.minor.patch` pattern, and it also states that implementations _should ignore_ the `.patch` part.

Which is quite confusing.

To avoid this confusion in the future, 

* #4233 

suggests to relax the pattern for the `openapi` field and also allow `major.minor` without a patch part.

**Comment by karenetheridge** [comment]
I'm going to continue this at https://github.com/OAI/OpenAPI-Specification/discussions/4233 as it's really not useful to have two places for the same topic.

**Comment by valerii15298** [comment]
Is there OpenAPI json schemas with other dialects than just `draft/2020-12` ?
I found this issue here: https://spec.openapis.org/#openapi-specification-schemas

And I hoped there will be info about openapi json schemas which use draft-07 schema, mainly because VS Code only supports draft-07.

<img width="590" height="33" alt="Image" src="https://github.com/user-attachments/assets/a4474987-a414-4377-892c-739dfc965e8a" />

As a side note can anyone suggest which editors support editing/exploring openapi json schema(basically ui editors that support `draft/2020-12`)? 

It is really difficult to work with json schema when almost no tools support its latest version... I know it is json schema related. But practically most people still use draft-07

**Comment by karenetheridge** [comment]
> And I hoped there will be info about openapi json schemas which use draft-07 schema, mainly because VS Code only supports draft-07.

Yes, you can certainly use draft7 in the schemas in your OpenAPI document -- you can do that with `jsonSchemaDialect`:

```yaml
openapi: 3.1.1
info:
  title: my API
  version: 1.0.0
jsonSchemaDialect: 'http://json-schema.org/draft-07/schema#'
...
```

I have no data on how often a document sets a custom dialect, but you can see which tools support draft7 and draft2020-12 here: https://json-schema.org/tools

**Comment by valerii15298** [comment]
As I see here: https://spec.openapis.org/oas/3.1/schema-base/2025-08-31.html the `jsonSchemaDialect` is set to const to be `"https://spec.openapis.org/oas/3.1/dialect/2024-11-10"`

But to clarify what I wanted is to have JSON Schema which describes syntax of openapi schema in draft-07.

On this page: https://spec.openapis.org/#openapi-specification-schemas json schemas which define openapi schema syntax all use JSON Schema versiion "https://json-schema.org/draft/2020-12/schema"

I am using this schema: https://spec.openapis.org/oas/3.1/schema/2025-08-31.html to define syntax for my openapi file, but vs code does not support it fully unfortunately...

**Comment by ralfhandl** [comment]
There are VS Code extensions for validating against 2020-12.

Haven't tried them because I typically use Ajv or https://www.npmjs.com/package/@hyperjump/json-schema in my test scripts to validate schemas and instance data.

**Comment by truongminhdien8330-web** [comment]
> Currently, we have the following schema document `$id`s for OAS 3.1 (shown relative to a base URI of `https://https://spec.openapis.org/oas/3.1/`, and without the trailing date segment that all schemas need to support updates:
> 
> * `schema`   # Validates everything _except_ the Schema Object, which is only validated against `type: [boolean, object]`
> * `schema-base`  # Extends `schema` to validate Schema Objects against `dialect/base` _and_ require both `$schema` and `jsonSchemaDialect` to be set to `schema`'s `$id` if they are present
> * `dialect/base` # A dialect equivalent to the default 2020-12 dialect plus the OAS extensions
> * `meta/base`  # The vocabulary metaschema for the OAS extensions
> 
> `schema-base` was named to indicate that it incorporates `dialect/base`.  No one is certain why the dialect and vocabulary are referred to as "base", but they are called that in the OAS text as well.  I vaguely recall discussing the dialect as a base for further OAS-relevant extensions?  Perhaps that is what the intent was.
> 
> AFAICT, the advantage of `schema` is that it's self-contained.  But giving it that name means that it is the schema that most people assume is what they want, yet it really does not produce the behavior they'd expect.
> 
> `schema-base` is closer to what people would expect, but is (in my view) overly restrictive in locking `$schema` to `dialect/base`, as it is valid to switch metaschemas if `$schema` appears alongside `$id`, and conforming JSON Schema implementations know to treat `$id`-containing subschemas as embedded resources and check for a changed metaschema.  However, this does ***not*** make `schema-base` wrong:  Users might indeed want to enforce using _only_ `dialect/base`, and this is the correct way to do that.
> 
> On the other hand (and contrary to some of my comments earlier on Slack), `schema-base` locking `jsonSchemaDialect` to `dialect/base` is probably correct.  If you want to set `jsonSchemaDialect` to something else, you would want a schema that validates Scheam Objects accordingly by default.
> 
> -----
> 
> We probably shouldn't change how `schema` behaves, but we should make it more obvious that it is incomplete.  We should also come up with some sort of intuitive name for a schema that is like `schema-base` but does not restrict `$schema`.  This name needs to convey that _most users will want this new schema, and **not** the plain_ `schema`.
> 
> Ideally, I'd prefer to make copies of the vocabulary and dialect and give them the URIs `dialect/oas` and `meta/oas` (`3.1` is already present earlier in the URI), but that is probably more trouble than it is worth given that I imagine very few people pay attention to those other than to just load them correctly when using `schema-base`.
> 
> The important thing is to have an intuitively-named schema that performs the default accepted Schema Object validation but still allows switching metaschemas on embedded resources.  That is the default behavior of the OAS, and it's not currently provided.

https://www.youtube.com/@satoshiNakamotolnstitute
